sev-types/src/command/mod.rs

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! The `command` module contains a number of types that are
+//! useful for interacting in a low-level SEV environment.
+
+use core::marker::PhantomData;
+
+use bitfield::bitfield;
+use bitflags::bitflags;
+
+use crate::platform::Version;
+
+/// Reset SEV persistent state.
+///
+/// (Chapter 5.5)
+#[repr(C, packed)]
+pub struct PlatformReset;
+
+// FIXME: https://github.com/rust-lang/rustfmt/issues/4085
+#[rustfmt::skip]
+bitflags! {
+    /// The platform's status flags.
+    #[derive(Default)]
+    pub struct PlatformStatusFlags: u8 {
+	/// If set, the platform is externally owned.
+	/// Else, it is self-owned (default state).
+        const OWNER = 1;
+    }
+}
+
+bitfield! {
+    /// Contains information that describes how the
+    /// platform is currently configured.
+    #[derive(Clone, Copy, Default, PartialEq, Eq)]
+    pub struct PlatformStatusConfig(u32);
+    impl Debug;
+
+    /// If set, SEV-ES is initialized for the platform.
+    pub encrypted_state, _: 0, 0;
+    reserved, _: 23, 1;
+
+    /// The firmware build ID for this API version.
+    pub build, _: 31, 24;
+}
+
+/// Query SEV platform status.
+///
+/// (Chapter 5.6; Table 17)
+#[derive(Default)]
+#[repr(C, packed)]
+pub struct PlatformStatus {
+    /// The firmware version (major.minor)
+    pub version: Version,
+
+    /// The Platform State.
+    pub state: u8,
+
+    /// Right now the only flag that is communicated in
+    /// this single byte is whether the platform is self-
+    /// owned or not. If the first bit is set then the
+    /// platform is externally owned. If it is cleared, then
+    /// the platform is self-owned. Self-owned is the default
+    /// state.
+    pub flags: PlatformStatusFlags,
+
+    /// Contains configuration information about the platform.
+    pub config: PlatformStatusConfig,
+
+    /// The number of valid guests maintained by the SEV firmware.
+    pub guest_count: u32,
+}
+
+/// Generate a new Platform Encryption Key (PEK).
+///
+/// (Chapter 5.7)
+#[repr(C, packed)]
+pub struct PekGenerate;
+
+/// Take ownership of the platform.
+///
+/// (Chapter 5.8)
+#[repr(C, packed)]
+pub struct PekCertificateSigningRequest<'a> {
+    /// The system's physical address of a memory region that
+    /// the platform will write the PEK certificate to.
+    addr: u64,
+
+    /// The length of the memory region (in bytes) that will contain
+    /// the PEK certificate.
+    len: u32,
+    _phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> PekCertificateSigningRequest<'a> {
+    #![allow(missing_docs)]
+    pub fn new(addr: u64, len: u32) -> Self {
+        Self {
+            addr,
+            len,
+            _phantom: PhantomData,
+        }
+    }
+}
```

Comment:
>     1. You've skipped documentation on the public function, but added documentation to the private fields. This should be reversed.

I'm unsure of what value a doc comment adds to that constructor function. I've been commenting the struct fields because I am trying to justify their layout by rooting it to its description in the AMD SEV API spec for easy cross-reference and confirmation of correctness.

>     2. This struct is not safe. You need to take a reference to something to bind the lifetime `'a`. I _think_ what we want is a `&'a mut Cert` instead of `addr: u64`, but I don't remember the details.

Okay. I might need to get some additional help on this. Here's what I'm envisioning what we want: we want the compiler to confirm that the struct's lifetime is enclosed by the lifetime of the memory address we're using here, which in our current code is cast all the way down to a u64. I don't know how to express this.